### ISSUE TITLE
CucumberException thrown from plugin

### ISSUE THREAD

**Issue by aslakhellesoy** [cucumber-ltd/cucumber-pro-plugin-jvm#17]
```
------------------------------------------------------- 
04-Dec-2017 10:12:00   T E S T S 
04-Dec-2017 10:12:00  ------------------------------------------------------- 
04-Dec-2017 10:12:00  Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=256m; support was removed in 8.0 
04-Dec-2017 10:12:02  Running com.vanguard.shared.advice.cucumber.RunIntegrationCukeTest 
04-Dec-2017 10:12:03  Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 1.075 sec <<< FAILURE! - in com.vanguard.shared.advice.cucumber.RunIntegrationCukeTest 
04-Dec-2017 10:12:03  com.vanguard.shared.advice.cucumber.RunIntegrationCukeTest  Time elapsed: 1.074 sec  <<< ERROR! 
04-Dec-2017 10:12:03  cucumber.runtime.CucumberException: java.lang.IllegalArgumentException: One of setGitDir or setWorkTree must be called. 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.instantiate(PluginFactory.java:114) 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.create(PluginFactory.java:87) 
04-Dec-2017 10:12:03          at cucumber.runtime.RuntimeOptions.getPlugins(RuntimeOptions.java:241) 
04-Dec-2017 10:12:03          at cucumber.runtime.RuntimeOptions$1.invoke(RuntimeOptions.java:287) 
04-Dec-2017 10:12:03          at com.sun.proxy.$Proxy17.uri(Unknown Source) 
04-Dec-2017 10:12:03          at cucumber.runtime.junit.JUnitReporter.uri(JUnitReporter.java:160) 
04-Dec-2017 10:12:03          at cucumber.runtime.junit.FeatureRunner.run(FeatureRunner.java:68) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.runChild(Cucumber.java:93) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.runChild(Cucumber.java:37) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner.run(ParentRunner.java:363) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.run(Cucumber.java:98) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103) 
04-Dec-2017 10:12:03  Caused by: java.lang.IllegalArgumentException: One of setGitDir or setWorkTree must be called. 
04-Dec-2017 10:12:03          at org.eclipse.jgit.lib.BaseRepositoryBuilder.requireGitDirOrWorkTree(BaseRepositoryBuilder.java:590) 
04-Dec-2017 10:12:03          at org.eclipse.jgit.lib.BaseRepositoryBuilder.setup(BaseRepositoryBuilder.java:555) 
04-Dec-2017 10:12:03          at org.eclipse.jgit.storage.file.FileRepositoryBuilder.build(FileRepositoryBuilder.java:92) 
04-Dec-2017 10:12:03          at io.cucumber.pro.revision.jgit.JGitRevisionProvider.<init>(JGitRevisionProvider.java:26) 
04-Dec-2017 10:12:03          at io.cucumber.pro.revision.jgit.JGitRevisionProvider.<init>(JGitRevisionProvider.java:16) 
04-Dec-2017 10:12:03          at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) 
04-Dec-2017 10:12:03          at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) 
04-Dec-2017 10:12:03          at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) 
04-Dec-2017 10:12:03          at java.lang.reflect.Constructor.newInstance(Constructor.java:423) 
04-Dec-2017 10:12:03          at java.lang.Class.newInstance(Class.java:442) 
04-Dec-2017 10:12:03          at io.cucumber.pro.revision.RevisionProviderFactory.create(RevisionProviderFactory.java:15) 
04-Dec-2017 10:12:03          at io.cucumber.pro.results.ResultsPublisherFactory.create(ResultsPublisherFactory.java:23) 
04-Dec-2017 10:12:03          at io.cucumber.pro.JsonReporter12.<init>(JsonReporter12.java:44) 
04-Dec-2017 10:12:03          at io.cucumber.pro.JsonReporter12.<init>(JsonReporter12.java:54) 
04-Dec-2017 10:12:03          at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) 
04-Dec-2017 10:12:03          at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) 
04-Dec-2017 10:12:03          at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) 
04-Dec-2017 10:12:03          at java.lang.reflect.Constructor.newInstance(Constructor.java:423) 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.instantiate(PluginFactory.java:107) 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.create(PluginFactory.java:87) 
04-Dec-2017 10:12:03          at cucumber.runtime.RuntimeOptions.getPlugins(RuntimeOptions.java:241) 
04-Dec-2017 10:12:03          at cucumber.runtime.RuntimeOptions$1.invoke(RuntimeOptions.java:287) 
04-Dec-2017 10:12:03          at com.sun.proxy.$Proxy17.uri(Unknown Source) 
04-Dec-2017 10:12:03          at cucumber.runtime.junit.JUnitReporter.uri(JUnitReporter.java:160) 
04-Dec-2017 10:12:03          at cucumber.runtime.junit.FeatureRunner.run(FeatureRunner.java:68) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.runChild(Cucumber.java:93) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.runChild(Cucumber.java:37) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268) 
04-Dec-2017 10:12:03          at org.junit.runners.ParentRunner.run(ParentRunner.java:363) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.run(Cucumber.java:98) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103) 
04-Dec-2017 10:12:03   
04-Dec-2017 10:12:03  com.vanguard.shared.advice.cucumber.RunIntegrationCukeTest  Time elapsed: 1.075 sec  <<< ERROR! 
04-Dec-2017 10:12:03  cucumber.runtime.CucumberException: Only one formatter can use STDOUT, now both pretty and pretty use it. If you use more than one formatter you must specify output path with PLUGIN:PATH_OR_URL 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.defaultOutOrFailIfAlreadyUsed(PluginFactory.java:185) 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.convertOrNull(PluginFactory.java:144) 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.instantiate(PluginFactory.java:99) 
04-Dec-2017 10:12:03          at cucumber.runtime.formatter.PluginFactory.create(PluginFactory.java:87) 
04-Dec-2017 10:12:03          at cucumber.runtime.RuntimeOptions.getPlugins(RuntimeOptions.java:241) 
04-Dec-2017 10:12:03          at cucumber.runtime.RuntimeOptions$1.invoke(RuntimeOptions.java:287) 
04-Dec-2017 10:12:03          at com.sun.proxy.$Proxy17.done(Unknown Source) 
04-Dec-2017 10:12:03          at cucumber.runtime.junit.JUnitReporter.done(JUnitReporter.java:208) 
04-Dec-2017 10:12:03          at cucumber.api.junit.Cucumber.run(Cucumber.java:99) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153) 
04-Dec-2017 10:12:03          at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128) 
04-Dec-2017 10:12:03   
04-Dec-2017 10:12:03   
04-Dec-2017 10:12:03  Results : 
04-Dec-2017 10:12:03   
04-Dec-2017 10:12:03  Tests in error:  
04-Dec-2017 10:12:03    RunIntegrationCukeTest.com.vanguard.shared.advice.cucumber.RunIntegrationCukeTest ª Cucumber 
04-Dec-2017 10:12:03    JUnit4Provider.invoke:128->executeTestSet:153->executeWithRerun:173->execute:283 ª Cucumber 
04-Dec-2017 10:12:03   
04-Dec-2017 10:12:03  Tests run: 2, Failures: 0, Errors: 2, Skipped: 0 
04-Dec-2017 10:12:03   
04-Dec-2017 10:12:03  [INFO] ------------------------------------------------------------------------ 
04-Dec-2017 10:12:03  [INFO] BUILD FAILURE 
04-Dec-2017 10:12:03  [INFO] ------------------------------------------------------------------------ 
04-Dec-2017 10:12:03  [INFO] Total time: 14.733 s 
04-Dec-2017 10:12:03  [INFO] Finished at: 2017-12-04T10:12:03-05:00 
04-Dec-2017 10:12:03  [INFO] Final Memory: 73M/1237M 
04-Dec-2017 10:12:03  [INFO] ------------------------------------------------------------------------ 
04-Dec-2017 10:12:03  [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.18.1:test (integration-cuke-tests) on project model-child-exemplar.webservice: There are test failures. 
04-Dec-2017 10:12:03  [ERROR]  
04-Dec-2017 10:12:03  [ERROR] Please refer to /opt/vgi/cto/bamboo/agents/bamboo-agent-6/xml-data/build-dir/ESSMCE-WEBSERVICE8-INTEGRATIONTEST/target/surefire-reports for the individual test results. 
04-Dec-2017 10:12:03  [ERROR] -> [Help 1] 
04-Dec-2017 10:12:03  [ERROR]  
04-Dec-2017 10:12:03  [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch. 
04-Dec-2017 10:12:03  [ERROR] Re-run Maven using the -X switch to enable full debug logging. 
04-Dec-2017 10:12:03  [ERROR]  
04-Dec-2017 10:12:03  [ERROR] For more information about the errors and possible solutions, please read the following articles: 
04-Dec-2017 10:12:03  [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException 
04-Dec-2017 10:12:03  Failing task since return code of [/opt/vgi/maven/apache-maven-3.2.5/bin/mvn prepare-package -DskipJasmineTests -U -Psanity-check,integration-tests -Djersey.config.test.container.port=59218 -DfailIfNoTests=false -Djava.io.tmpdir=/opt/vgi/cto/bamboo/tmp -s /opt/vgi/maven/apache-maven-3.2.5/conf/settings.xml -DJava_Home=/opt/vgi/java/jdk1.8.0_current] was 1 while expected 0 
04-Dec-2017 10:12:03  Finished task 'Runs tests categorized as integration tests' with result: Failed 
04-Dec-2017 10:12:03  Running post build plugin 'NCover Results Collector' 
04-Dec-2017 10:12:03  Running post build plugin 'Clover Results Collector' 
04-Dec-2017 10:12:03  Running post build plugin 'Docker Container Cleanup' 
04-Dec-2017 10:12:03  Running post build plugin 'npm Cache Cleanup' 
04-Dec-2017 10:12:03  Running post build plugin 'Artifact Copier' 
04-Dec-2017 10:12:03  Publishing an artifact: SurefireIntegrationTests 
04-Dec-2017 10:12:03  Finished publishing of artifact Shared artifact: [SurefireIntegrationTests], pattern: [target/surefire-reports/*integration*] anchored at: [.] in 15.05 ms 
04-Dec-2017 10:12:03  Publishing an artifact: Classes 
04-Dec-2017 10:12:03  Finished publishing of artifact Shared artifact: [Classes], pattern: [target/classes/**] anchored at: [.] in 44.57 ms 
04-Dec-2017 10:12:03  Publishing an artifact: CucumberIntegrationTests 
04-Dec-2017 10:12:03  Finished publishing of artifact Shared artifact: [CucumberIntegrationTests], pattern: [target/cucumber-reports/integration/**] anchored at: [.] in 9.427 ms 
04-Dec-2017 10:12:03  Publishing an artifact: JacocoIntegrationTests 
04-Dec-2017 10:12:03  Finished publishing of artifact Shared artifact: [JacocoIntegrationTests], pattern: [target/jacoco-it.exec] anchored at: [.] in 9.769 ms 
04-Dec-2017 10:12:04  Successfully removed working directory at '/opt/vgi/cto/bamboo/agents/bamboo-agent-6/xml-data/build-dir/ESSMCE-WEBSERVICE8-INTEGRATIONTEST' 
04-Dec-2017 10:12:04  Finalising the build... 
04-Dec-2017 10:12:04  Stopping timer. 
04-Dec-2017 10:12:04  Build ESSMCE-WEBSERVICE8-INTEGRATIONTEST-5 completed. 
04-Dec-2017 10:12:04  Running on server: post build plugin 'Build Hanging Detection Configuration' 
04-Dec-2017 10:12:04  Running on server: post build plugin 'Clover Delta Calculator' 
04-Dec-2017 10:12:04  Running on server: post build plugin 'Maven Dependencies Postprocessor' 
04-Dec-2017 10:12:04  All post build plugins have finished 
04-Dec-2017 10:12:04  Generating build results summary... 
04-Dec-2017 10:12:04  Saving build results to disk... 
04-Dec-2017 10:12:04  Logging substituted variables... 
04-Dec-2017 10:12:04  Indexing build results... 
04-Dec-2017 10:12:04  Finished building ESSMCE-WEBSERVICE8-INTEGRATIONTEST-5. 
```

@tooky can you confirm what plugin version this is for?

**Comment by aslakhellesoy** [comment]
And Cucumber-JVM as well as Java version please

**Comment by tooky** [comment]
@aslakhellesoy it was for 1.2.0. 

We saw it both with an older Cucumber (1.2.6?) and with 2.1.0. We upgraded to see if it fixed it. 

Unsure about java version, will follow up. 

It was running on a bamboo CI env.

**Comment by aslakhellesoy** [comment]
Has anyone on this team or this organisation successfully run the plugin before?

**Comment by aslakhellesoy** [comment]
One possible reason for this error is that the current directory (or any of its parents) doesn't have a `.git` directory, so JGit isn't able to detect where the repo root is.

**Comment by tooky** [comment]
No, we are trying to get it set up for the first time. We have been able to publish documentation using curl in the past. I will check where that is happening from.

**Comment by mattwynne** [comment]
There is no `.git` directory on their CI build agents IIRC. That’s why we had to use a bamboo env var for the git rev when we used curl, as those git commands didn’t work.

**Comment by tooky** [comment]
I meant git over ssh, how do we do that?

**Comment by aslakhellesoy** [comment]
> There is no .git directory on their CI build agents IIRC

That's odd. How does Bamboo fetch the latest code before building then? `git pull && rm -rf .git`? Why?

Currently we're getting the current revision by consulting git (using jgit). If there is no `.git` directory (weird) that will obviously not work. It would be trivial to add a different strategy to get the revision from an env var instead.

*However* - if there is no `.git` directory the plugin won't be able to publish documentation at all (`git push`), so this workaround doesn't sound like it would work in practice. Cucumber Pro can't do anything meaningful with results if it doesn't also have the files.

**Comment by aslakhellesoy** [comment]
> We have been able to publish documentation using curl in the past.

Last time I checked the only way to publish *documentation* is via `git push`.
Only *results* can be pushed with HTTP (`curl`).

Is that what you mean't @tooky?

**Comment by aslakhellesoy** [comment]
Here is a [guide](https://confluence.atlassian.com/bamkb/how-to-commit-to-a-git-repo-during-a-build-720641378.html) explaining how to access the local git repo during the build.

Perhaps you're dealing with a team that doesn't use git, but Subversion?

**Comment by tooky** [comment]
It's definitely git.

**Comment by aslakhellesoy** [comment]
@tooky @mattwynne here is how to debug this further:

* Upgrade to [1.2.1](https://github.com/cucumber-ltd/cucumber-pro-plugin-jvm/blob/master/CHANGELOG.md#121---2017-12-04)
* Define `CUCUMBER_PRO_LOG_LEVEL=DEBUG`

This should give more detailed logging, which would help us diagnose the root cause of this.

Again - I don't believe the build slaves don't have git, for two reasons:
* The CI server needs git to clone/fetch the project it's building
* You have previously published documentation with a manual `git push`

**Comment by aslakhellesoy** [comment]
To make progress on this we need to know:

* Do the slaves' working copy of a project have a `.git` directory?
* Can you log into the slave, cd to a working copy and run `git status` successfully (i.e. is it uncorrupted)?
* What is the working directory when a build slave runs a build. Is it *different* from the working copy directory?
  * To find this out you may have to run a build that can print the current directory to STDOUT during build. In Java: `System.out.println("WORKING DIR: " + System.getProperty("user.dir"))` 

It would also be useful to know:

* Is `git` installed on slaves?
* Is it on the `$PATH` - if not, where is it?

**Comment by aslakhellesoy** [comment]
@tooky @mattwynne have you managed to find any answers to the above?

**Comment by aslakhellesoy** [comment]
We should probably just throw an error explaining that there is no git working copy.

**Comment by aslakhellesoy** [comment]
Fixed in 3f147628c3ba1bb9cde643c94a436be7937b2216
